File: MATLAB/Method/Statistical_methods/Frequency_domain/WT2/WT2.m
function [Pulse_shape_discrimination_factor] = WT2(Pulse_signal, Sigma, Wavelet_length, Dt)

%--------------------------------------------------------------------------
% Performs pulse shape discrimination using Marr Wavelet Transform (WT2).
% 
% Uses Mexican Hat (Marr) wavelet transform to analyze signal features.
% Discrimination factor is computed as ratio between original signal integral
% and positive parts of the wavelet-transformed signal.
%
% Inputs:
%   Sigma: Wavelet width parameter (default: sqrt(5))
%   Wavelet_length: Half-width of wavelet support (default: 100)
%   Dt: Sampling time interval (default: 1.0)
%
% Reference:
% Langeveld, Willem GJ, et al. "Pulse shape discrimination algorithms, 
% figures of merit, and gamma-rejection for liquid and solid scintillators."
% IEEE Transactions on Nuclear Science 64.7 (2017): 1801-1809.
%--------------------------------------------------------------------------
    
    % Set default parameters
    if nargin < 2
        Sigma = sqrt(5);        % Controls wavelet spread
    end
    if nargin < 3
        Wavelet_length = 100;   % Controls analysis window
    end
    if nargin < 4
        Dt = 1.0;              % Time step between samples
    end
    
    % Initialize arrays
    Num_signals = size(Pulse_signal, 1);
    Pulse_shape_discrimination_factor = zeros(Num_signals, 1);
    
    % Generate Marr (Mexican Hat) wavelet
    T_wavelet = linspace(-Wavelet_length, Wavelet_length, 2 * Wavelet_length + 1);
    Wavelet = marr_wavelet(T_wavelet, Sigma);
    
    % Process each signal
    for i = 1:Num_signals
        % Get current signal
        Signal = Pulse_signal(i, :);
        
        % Create time vector for integration
        Time_vector = (0:length(Signal)-1) * Dt;
        
        % Calculate original signal integral
        Integral_signal = trapz(Time_vector, Signal);
        
        % Apply wavelet transform via convolution
        Convolved_signal = conv(Signal, Wavelet, 'same');
        
        % Calculate integral of positive wavelet components
        Positive_convolved = max(0, Convolved_signal);
        Integral_wavelet = trapz(Time_vector, Positive_convolved);
        
        % Calculate discrimination factor:
        % 2 * Original_Integral / (Original_Integral + Wavelet_Integral)
        Pulse_shape_discrimination_factor(i) = 2 * Integral_signal / (Integral_signal + Integral_wavelet);
    end
end

function Wavelet = marr_wavelet(t, Sigma)
    %--------------------------------------------------------------------------
    % Generates Marr (Mexican Hat) wavelet.
    %
    % Formula: ψ(t) = (1 - t²/σ²) * exp(-t²/(2σ²))
    %
    % Inputs:
    %   t: Time points for wavelet evaluation
    %   Sigma: Width parameter controlling spread
    %--------------------------------------------------------------------------
    Wavelet = (1 - (t.^2) / (Sigma^2)) .* exp(-t.^2 / (2 * Sigma^2));
end

File: Method/Statistical_methods/Frequency_domain/wt2.py
"""
Wavelet Transform (WT2) method with Marr wavelet for pulse shape discrimination.

This module implements a wavelet-based approach to pulse shape discrimination
using the Marr (Mexican Hat) wavelet transform. The discrimination factor is
computed as a ratio of signal integrals in different domains.

Reference:
- Langeveld, Willem GJ, Michael J. King, John Kwong, and Daniel T. Wakeford.
  "Pulse shape discrimination algorithms, figures of merit, and gamma-rejection for liquid and solid scintillators."
  IEEE Transactions on Nuclear Science 64, no. 7 (2017): 1801-1809.
"""

import numpy as np
from scipy.signal import convolve
from numpy import trapz

def marr_wavelet(t: np.ndarray, sigma: float = 1.0) -> np.ndarray:
    """
    Compute the Marr (Mexican Hat) wavelet function values.

    The Marr wavelet is defined as:
        ψ(t) = (1 - t^2/sigma^2) * exp(-t^2/(2*sigma^2))

    Args:
        t: 1D numpy array of time values
        sigma: Standard deviation of the wavelet (default: 1.0)

    Returns:
        1D numpy array of Marr wavelet values
    """
    return (1 - (t ** 2) / (sigma ** 2)) * np.exp(-t ** 2 / (2 * sigma ** 2))


def get_psd_factor(
    pulse_signal: np.ndarray,
    sigma: float = np.sqrt(5),
    wavelet_length: int = 100,
    dt: float = 1.0
) -> np.ndarray:
    """
    Performs pulse shape discrimination based on the Wavelet Transform (WT).

    WT2 computes the ratio of the integral of the signal in the frequency domain obtained through 
    the Marr wavelet to the integral that best reflects the difference in signal characteristics, 
    which serves as the discrimination factor.

    Args:
        pulse_signal: A 2D numpy array where each row corresponds to a pulse signal
        sigma: Standard deviation for the Marr wavelet (default: sqrt(5))
        wavelet_length: Half-length of the wavelet; the wavelet is computed over 
                       [-wavelet_length, wavelet_length] (default: 100)
        dt: Time interval between samples (default: 1.0)

    Returns:
        A numpy array of discrimination factors
    """
    num_signals = pulse_signal.shape[0]
    discrimination_factors = np.zeros(num_signals)  # Initialize the array to store PSD values

    # Define time vector for the wavelet and compute the Marr wavelet
    t_wavelet = np.linspace(-wavelet_length, wavelet_length, 2 * wavelet_length + 1)
    wavelet = marr_wavelet(t_wavelet, sigma)

    for i in range(num_signals):
        # Extract the current pulse signal
        signal = pulse_signal[i, :]

        # Create a time vector for integration (assuming dt spacing)
        time_vector = np.arange(0, len(signal)) * dt

        # Compute the integral of the original signal
        integral_signal = trapz(signal, time_vector)

        # Perform convolution of the signal with the Marr wavelet
        convolved_signal = convolve(signal, wavelet, mode='same')

        # Compute the integral of the positive part of the wavelet-transformed signal
        positive_convolved = np.maximum(0, convolved_signal)
        integral_wavelet = trapz(positive_convolved, time_vector)

        # Calculate the PSD value and store it in the array
        discrimination_factors[i] = 2 * integral_signal / (integral_signal + integral_wavelet)

    return discrimination_factors
